mathics/eval/tensors.py

```diff
@@ -156,74 +233,51 @@ def eval_Outer(f, lists, evaluation: Evaluation):
     if sparse_to_list:
         lists = new_lists
 
-    def rec(item, rest_lists, current):
-        evaluation.check_stopped()
-        if isinstance(item, Atom) or not item.head.sameQ(head):
-            if rest_lists:
-                return rec(rest_lists[0], rest_lists[1:], current + [item])
-            else:
-                return Expression(f, *(current + [item]))
-        else:
-            elements = []
-            for element in item.elements:
-                elements.append(rec(element, rest_lists, current))
-            return Expression(head, *elements)
-
-    def rec_sparse(item, rest_lists, current):
-        evaluation.check_stopped()
-        if isinstance(item, tuple):  # (rules)
-            elements = []
-            for element in item:
-                elements.extend(rec_sparse(element, rest_lists, current))
-            return tuple(elements)
-        else:  # rule
-            _pos, _val = item.elements
-            if rest_lists:
-                return rec_sparse(
-                    rest_lists[0],
-                    rest_lists[1:],
-                    (current[0] + _pos.elements, current[1] * _val),
-                )
-            else:
-                return (
-                    Expression(
-                        SymbolRule,
-                        ListExpression(*(current[0] + _pos.elements)),
-                        current[1] * _val,
-                    ),
-                )
-
     # head != SparseArray
     if not head.sameQ(SymbolSparseArray):
-        return rec(lists[0], lists[1:], [])
+        etc = (
+            (lambda item, level: isinstance(item, Atom) or not item.head.sameQ(head)),

```

Comment:
Many linters frown on using lambda and prefer `def` functions instead.

Here this would look like;

```python
            def sparse_get_item_fn(item, level):
	    	return: isinstance(item, Atom) or not item.head.sameQ(head)
	    def sparse_get_elements_fn(head, *elements)
```

As before, there could be a whole topic on the wisdom of deprecating lambda's which personally I find questionable other than the crappy irregular syntax of `lambda` that was a mistake made from a long time ago and not corrected.

I am just reporting what Python linters report and what my understanding of Python convention is.

Reply:
As a frequent mathematica user I'm used to pure functions. Given that there have been hundreds of ``lambda`` in Mathic3, I don't feel it necessary to rewrite pure functions like ``lambda elements: elements``, which is clear and can be written in a single line, with ``def``. However, I agree that rewriting longer pure functions with ``def`` is helpful. Now all pure functions longer than one line is replaced with ``def``.